File: richos/tools/richos-service/companion-windows/tests/RichOSCompanionCore.Tests/ChannelMixerTests.cs
using RichOSCompanionCore;
using Xunit;

namespace RichOSCompanionCore.Tests
{
    /// <summary>
    /// The frozen audio invariant is load-bearing: LEFT = me (mic), RIGHT = others (system). A swapped
    /// L/R here would silently mis-attribute every speaker in every transcript — so it is tested to
    /// death, exactly like the macOS companion's <c>ChannelMixerTests.swift</c>.
    /// </summary>
    public class ChannelMixerTests
    {
        [Fact]
        public void FloatToInt16ScalesAndClamps()
        {
            Assert.Equal(0, ChannelMixer.FloatToInt16(0f));
            Assert.Equal((short)32767, ChannelMixer.FloatToInt16(1.0f));
            Assert.Equal((short)-32767, ChannelMixer.FloatToInt16(-1.0f));
            Assert.Equal((short)32767, ChannelMixer.FloatToInt16(2.0f));   // clamp high
            Assert.Equal((short)-32767, ChannelMixer.FloatToInt16(-2.0f)); // clamp low
        }

        [Fact]
        public void InterleavePutsMicOnLeftSystemOnRight()
        {
            var samples = ChannelMixer.InterleaveToInt16(new[] { 1.0f, 0f }, new[] { -1.0f, 0f });
            // frame 0: LEFT=mic(+1)->32767, RIGHT=system(-1)->-32767
            Assert.Equal((short)32767, samples[0]);
            Assert.Equal((short)-32767, samples[1]);
            Assert.Equal((short)0, samples[2]);
            Assert.Equal((short)0, samples[3]);
        }

        [Fact]
        public void Int16LeBytesAreLittleEndian()
        {
            var bytes = ChannelMixer.Int16LeBytes(new short[] { 0x0102, unchecked((short)0xFFFE) });
            Assert.Equal(new byte[] { 0x02, 0x01, 0xFE, 0xFF }, bytes);
        }

        [Fact]
        public void DownmixAveragesChannels()
        {
            // stereo interleaved: frame0 (1.0, 0.0), frame1 (0.0, 1.0) -> mono 0.5, 0.5
            var mono = ChannelMixer.DownmixToMono(new[] { 1.0f, 0.0f, 0.0f, 1.0f }, 2);
            Assert.Equal(2, mono.Length);
            Assert.Equal(0.5f, mono[0], 5);
            Assert.Equal(0.5f, mono[1], 5);
        }

        [Fact]
        public void PeakAndRms()
        {
            Assert.Equal(0.8, ChannelMixer.Peak(new[] { -0.8f, 0.2f }), 5);
            Assert.Equal(0.5, ChannelMixer.Rms(new[] { 0.5f, -0.5f }), 5);
        }
    }
}

File: richos/tools/richos-service/companion-macos/Tests/RichOSCompanionCoreTests/ChannelMixerTests.swift
import XCTest
@testable import RichOSCompanionCore

/// The frozen L=mic / R=system invariant is the one thing that, if wrong, silently mis-attributes
/// every speaker. These tests pin it mechanically.
final class ChannelMixerTests: XCTestCase {

    func testFloatToInt16ClampsAndScales() {
        XCTAssertEqual(ChannelMixer.floatToInt16(0), 0)
        XCTAssertEqual(ChannelMixer.floatToInt16(1.0), 32767)
        XCTAssertEqual(ChannelMixer.floatToInt16(-1.0), -32767)
        XCTAssertEqual(ChannelMixer.floatToInt16(2.0), 32767)   // clamps high
        XCTAssertEqual(ChannelMixer.floatToInt16(-2.0), -32767) // clamps low
    }

    func testInterleaveMapsMicToLeftSystemToRight() {
        // Distinct amplitudes so L/R can't be confused.
        let mic: [Float] = [1.0, 0.5]
        let sys: [Float] = [-1.0, -0.5]
        let out = ChannelMixer.interleaveToInt16(mic: mic, system: sys)
        // frame 0: L=mic(1.0)=32767, R=sys(-1.0)=-32767 ; frame 1: L=0.5, R=-0.5
        XCTAssertEqual(out[0], 32767)          // LEFT  = me
        XCTAssertEqual(out[1], -32767)         // RIGHT = others
        XCTAssertEqual(out[2], ChannelMixer.floatToInt16(0.5))
        XCTAssertEqual(out[3], ChannelMixer.floatToInt16(-0.5))
        XCTAssertEqual(out.count, 4)
    }

    func testDownmixAveragesChannels() {
        // interleaved stereo: [L0,R0, L1,R1]
        let interleaved: [Float] = [1.0, 0.0, 0.0, 1.0]
        let mono = ChannelMixer.downmixToMono(interleaved: interleaved, channelCount: 2)
        XCTAssertEqual(mono, [0.5, 0.5])
    }

    func testInt16LEByteOrder() {
        let bytes = ChannelMixer.int16LEBytes([0x0102])
        XCTAssertEqual(bytes, [0x02, 0x01]) // little-endian
    }

    func testPeakAndRms() {
        XCTAssertEqual(ChannelMixer.peak([-0.3, 0.7, -0.2]), 0.7, accuracy: 1e-6)
        XCTAssertEqual(ChannelMixer.rms([1.0, -1.0]), 1.0, accuracy: 1e-6)
        XCTAssertEqual(ChannelMixer.rms([]), 0.0, accuracy: 1e-6)
    }
}
